add doc for third party ingress (#13116)
istio/istio.io: content/en/docs/ops/integrations/loadbalancers/index.md

---
title: Third Party Load Balancers
description: How to integrate Istio with third party load balancers.
weight: 90
keywords: [traffic-management,ingress]
owner: istio/wg-networking-maintainers

> test: no

Reviewer: ```suggestion
test: n/a
```

---

Istio provides both an ingress and service mesh implementation, which can be used together or separately. While these are designed to work together seamlessly, there are times when integrating with a third party ingress is required. This could be for migration purposes, feature requirements, or personal preferences.

## Integration Modes

In "standalone" mode, the third party ingress is directly sending to backends.

> In this case, the backends presumably have Istio sidecars injected;

Reviewer: ```suggestion
In this case, the backends presumably have Istio sidecars injected.
```

{{< mermaid >}}
graph LR
    cc((Client))
    tpi(Third Party Ingress)
    a(Backend)
    cc-->tpi-->a
{{< /mermaid >}}

In this mode, things mostly just work.

> Clients in a service mesh do not need to be aware that the backed they are connecting to have a sidecar.

Reviewer 2: ```suggestion
Clients in a service mesh do not need to be aware that the backed they are connecting to has a sidecar.
```

Reviewer: also backed -> backend
```suggestion
Clients in a service mesh do not need to be aware that the backend they are connecting to has a sidecar.
```

However, the ingress will not use mTLS, which may lead to undesirable behavior.
As a result, most of the configuration for this setup is around enabling mTLS.

> In "chained" mode, we use both the third party ingress *and* Istio's own Gateway in parallel.

Reviewer 2: I'm not sure I like the word `parallel` here, as it means something different in a circuit and are really in `series`. Not sure I have a bite wording, and maybe just leave it off:
```suggestion
In "chained" mode, we use both the third party ingress *and* Istio's own Gateway.
```

Reviewer: Maybe `in parallel` -> `in sequence` ?

This can be useful when you want the functionality of both layers.

> In particular, this is useful with managed cloud load balancers, which have features like global addresses and managed certificates

Reviewer 2: ```suggestion
In particular, this is useful with managed cloud load balancers, which have features like global addresses and managed certificates.
```

{{< mermaid >}}
graph LR
    cc((Client))
    tpi(Third Party Ingress)
    ii(Istio Gateway)
    a(Backend)
    cc-->tpi
    tpi-->ii
    ii-->a
{{< /mermaid >}}

## Cloud Load Balancers

Generally, cloud load balancers will work out of the box in standalone mode without mTLS.
Vendor specific configuration is required to support chained mode or standalone with mTLS.

> ### Google HTTP(S) Load Balancer

Reviewer 2: @frankbu may have other thoughts, but I think we try to keep the documents platform neutral so not sure if this section should be here.

Reviewer 2: OK. We would just expect Any other providers who wanted to provide info to submit a PR to add it here. I'm not sure if that warrants a note or not, but I don't really see any notes elsewhere in the doc.

> Integration with Google HTTP(S) Load Balancers works out of the box with standalone mode if mTLS is not required.

Reviewer: ```suggestion
Integration with Google HTTP(S) Load Balancers only works out of the box with standalone mode if mTLS is not required as
```

mTLS is not supported.

Chained mode is possible. See [Google documentation](https://cloud.google.com/architecture/exposing-service-mesh-apps-through-gke-ingress) for setup instructions.

## In-Cluster Load Balancers

Generally, in-cluster load balancers will work out of the box in standalone mode without mTLS.

Standalone mode with mTLS can be achieved by inserting a sidecar into the Pod of the in-cluster load balancer.
This typically involves two steps beyond standard sidecar injection:

1. Disable inbound traffic redirection.
   While not required, typically we only want to use the sidecar for *outbound* traffic - inbound connection from clients is already handled by the load balancer itself.
   This also allows preserving the original client IP address, which would otherwise be lost by the sidecar.
   This mode can be enabled by inserting the `traffic.sidecar.istio.io/includeInboundPorts: ""` annotation on the load balancer `Pod`s.
1. Enable Service routing.

>    Istio sidecars can only properly function when requests are sent to Service's, not to specific pod IPs.

Reviewer 2: ```suggestion
   Istio sidecars can only properly function when requests are sent to Services, not to specific pod IPs.
```
